### PR TITLE
Support enableServiceMonitor in operator

### PR DESCRIPTION
**Pull request by clyang82**
support configure `serviceMonitor` in operator.

### CODE REVIEW

**Comment by clyang82** [comment]
@squat @kakkoyun Would you have time to re-review this? Thanks.

**Comment by 4n4nd** [comment]
@clyang82 would it make sense to add this for LokiSpec as well?

**Comment by clyang82** [comment]
> @clyang82 would it make sense to add this for LokiSpec as well?

Yes. That makes sense. I need to fix the build and test issue firstly and then update LokiSpec.

**Comment by clyang82** [comment]
@squat still need your help to review. Thank you.

File: api/v1alpha1/observatorium_types.go [hunk 1]
```diff
@@ -104,6 +104,9 @@ type ThanosReceiveControllerSpec struct {
 	// Compute Resources required by this container.
 	// +optional
 	Resources v1.ResourceRequirements `json:"resources,omitempty"`
+	// ServiceMonitor enables servicemonitor.
```
Comment (squat) [inline api/v1alpha1/observatorium_types.go:107]:
```suggestion
	// ServiceMonitor enables deploying a service monitor for the Thanos Receiver Controller.
```

File: api/v1alpha1/observatorium_types.go [hunk 2]
```diff
@@ -120,6 +123,9 @@ type ReceiversSpec struct {
...
```
Comment (squat) [inline api/v1alpha1/observatorium_types.go:126]:
```suggestion
	// ServiceMonitor enables deploying a service monitor for the Thanos Receivers.
```

File: api/v1alpha1/observatorium_types.go [hunk 3]
```diff
@@ -135,6 +141,9 @@ type StoreSpec struct {
...
```
Comment (squat) [inline api/v1alpha1/observatorium_types.go:144]:
```suggestion
	// ServiceMonitor enables deploying a service monitor for the Thanos Stores.
```

File: api/v1alpha1/observatorium_types.go [hunk 4]
```diff
@@ -157,6 +166,9 @@ type StoreCacheSpec struct {
 	// Compute Resources required by this container.
 	// +optional
 	ExporterResources v1.ResourceRequirements `json:"exporterResources,omitempty"`
+	// ServiceMonitor enables servicemonitor.
```
Comment (squat) [inline api/v1alpha1/observatorium_types.go:169]:
```suggestion
	// ServiceMonitor enables deploying a service monitor for the Thanos Store Caches.
```

File: api/v1alpha1/observatorium_types.go [hunk 5]
```diff
@@ -277,6 +289,9 @@ type APISpec struct {
 	// Compute Resources required by this container.
 	// +optional
 	Resources v1.ResourceRequirements `json:"resources,omitempty"`
+	// ServiceMonitor enables servicemonitor.
```
Comment (squat) [inline api/v1alpha1/observatorium_types.go:292]:
```suggestion
	// ServiceMonitor enables deploying a service monitor for the Observatorium API.
```

File: api/v1alpha1/observatorium_types.go [hunk 6]
```diff
@@ -289,6 +304,9 @@ type QuerySpec struct {
...
```
Comment (squat) [inline api/v1alpha1/observatorium_types.go:307]:
```suggestion
	// ServiceMonitor enables deploying a service monitor for the Thanos Queriers.
```

File: api/v1alpha1/observatorium_types.go [hunk 7]
```diff
@@ -322,6 +340,9 @@ type RuleSpec struct {
 	// Compute Resources required by this container.
 	// +optional
 	ReloaderResources v1.ResourceRequirements `json:"reloaderResources,omitempty"`
+	// ServiceMonitor enables servicemonitor.
```
Comment (squat) [inline api/v1alpha1/observatorium_types.go:343]:
```suggestion
	// ServiceMonitor enables deploying a service monitor for the Thanos Rulers.
```

File: api/v1alpha1/observatorium_types.go [hunk 8]
```diff
@@ -344,6 +365,9 @@ type CompactSpec struct {
 	// Compute Resources required by this container.
 	// +optional
 	Resources v1.ResourceRequirements `json:"resources,omitempty"`
+	// ServiceMonitor enables servicemonitor.
```
Comment (squat) [inline api/v1alpha1/observatorium_types.go:368]:
```suggestion
	// ServiceMonitor enables deploying a service monitor for the Thanos Compactors.
```

File: api/v1alpha1/observatorium_types.go [hunk 9]
```diff
@@ -360,6 +384,9 @@ type QueryFrontendSpec struct {
...
```
Comment (squat) [inline api/v1alpha1/observatorium_types.go:387]:
```suggestion
	// ServiceMonitor enables deploying a service monitor for the Thanos Query Frontends.
```

File: api/v1alpha1/observatorium_types.go [hunk 10]
```diff
@@ -378,6 +405,9 @@ type LokiSpec struct {
 	Version string `json:"version,omitempty"`
 	// VolumeClaimTemplate
 	VolumeClaimTemplate VolumeClaimTemplate `json:"volumeClaimTemplate"`
+	// ServiceMonitor enables servicemonitor.
```
Comment (squat) [inline api/v1alpha1/observatorium_types.go:408]:
```suggestion
	// ServiceMonitor enables deploying a service monitor for Loki.
```

**Review by squat — CHANGES_REQUESTED** [review]
Let's make the comments full sentences so they serve as meaningful documentation for users.

File: manifests/crds/core.observatorium.io_observatoria.yaml [hunk 11]
```diff
@@ -1321,6 +1329,10 @@ spec:
                       type: integer
                     description: Loki replicas per component
                     type: object
+                  serviceMonitor:
+                    description: ServiceMonitor enables  deploying a service monitor
```
Comment (squat) [inline manifests/crds/core.observatorium.io_observatoria.yaml:1333]: We need to regenerate this CRD now. Funny, i think CI should really catch that this file is out of date

**Review by squat — CHANGES_REQUESTED** [review]
Thanks for making the changes, @clyang82 . We need to regenerate the CRD and then let's merge

File: manifests/crds/core.observatorium.io_observatoria.yaml (hunk 11 above)
Comment (clyang82) [reply manifests/crds/core.observatorium.io_observatoria.yaml:1333]: I regenerated this CRD. That is why it can pass CI.

Comment (clyang82) [reply manifests/crds/core.observatorium.io_observatoria.yaml:1333]: I found there is an redundant space here. Let me update it.

**Review by squat — APPROVED** [review]

**Pull request merged**